### PR TITLE
Fix `implicit-function-declaration` warnings, passes 1..3

### PR DESCRIPTION
**Pull request by ribbanya**

### CODE REVIEW

File: src/placeholder.h [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#ifndef PLACEHOLDER_H
+#define PLACEHOLDER_H
+
+#include <dolphin/os/os.h>
+
+#define NOT_IMPLEMENTED(name)                                                  \
+    OSPanic(__FILE__, __LINE__, "#name is not implemented!");
```
Comment (ribbanya) [inline src/placeholder.h:7]: @r-burns Do you think `__func__` would work instead of `name`? I don't know how standard it is or if there are any gotchas.

Comment (r-burns) [reply src/placeholder.h:7]: Yeah I think that would work great. You just have to be careful as it's not really a macro https://stackoverflow.com/questions/31430765/why-func-function-and-pretty-function-arent-preprocessor-macros and detecting support is a little different https://stackoverflow.com/questions/15305310/predefined-macros-for-function-name-func

Comment (ribbanya) [reply src/placeholder.h:7]: So like this?
```suggestion
    OSPanic(__FILE__, __LINE__, "%s is not implemented!", __func__);
```

Comment (r-burns) [reply src/placeholder.h:7]: Yeah exactly

File: src/dolphin/card/CARDCreate.c [hunk 2]
```diff
@@ -112,9 +122,12 @@ s32 CARDCreateAsync(s32 chan, const char* fileName, u32 size,
     fileInfo->chan = chan;
     fileInfo->fileNo = freeNo;
 
-    result = __CARDAllocBlock(chan, size / card->sectorSize, CreateCallbackFat);
-    if (result < 0) {
+    /// @todo Eliminate cast to #CARDCallback.
+    result = __CARDAllocBlock(chan, size / card->sectorSize,
+                              (CARDCallback) CreateCallbackFat);
```
Comment (r-burns) [inline src/dolphin/card/CARDCreate.c:127]: Seems like you can remove these CARDCallback casts if you keep propagating EXIChannel to these function definitions

File: src/dolphin/os/OSAlarm.c [hunk 3]
```diff
@@ -24,9 +24,10 @@ void OSInitAlarm(void)
     }
 }
 
-void OSCreateAlarm(s32* arg0)
+void OSCreateAlarm(OSAlarm* alarm)
 {
-    *arg0 = 0;
+    /// @todo Hack.
+    *((s32*) alarm) = 0;
```
Comment (r-burns) [inline src/dolphin/os/OSAlarm.c:30]:
```suggestion
    alarm->handler = NULL;
```

File: src/dolphin/os/OSExi.c [hunk 4]
```diff
@@ -111,17 +114,21 @@ BOOL EXIImmEx(EXIChannel chan, void* buf, s32 len, u32 mode)
...
 
-        if (!EXISync(chan)) {
+        if (!EXISync(chan))
             return FALSE;
-        }
 
+/// @todo what
+#if defined(__MWERKS__) && defined(GEKKO)
         (u8*) buf += xLen;
+#endif
```
Comment (r-burns) [inline src/dolphin/os/OSExi.c:127]:
```suggestion
        buf = (u8*) buf + xLen;
```

File: src/melee/ft/chara/ftCLink/ftclink.h [hunk 5]
```diff
@@ -4,6 +4,7 @@
 #include <melee/ft/fighter.h>
 #include <sysdolphin/baselib/gobj.h>
 
+/// @todo this is #ftLinkAttributes.
```
Comment (r-burns) [inline src/melee/ft/chara/ftCLink/ftclink.h:7]: do you mean ftCLinkAttributes? Or are they the same for both

File: include/string.h [hunk 6]
```diff
@@ -3,7 +3,6 @@
 
 #include <stddef.h>
 
-void memcpy(void* dst, void* src, size_t num);
```
Comment (r-burns) [inline include/string.h:6]: Why'd this get deleted?

**Review by r-burns — COMMENTED** [review]
+1 to your plans to remove all the redundant typedef aliases and delete include/.

File: include/dolphin/types.h [hunk 7]
```diff
@@ -64,11 +64,17 @@ typedef jmp_t jtbl_t[];
...
+
 #elif defined(__MWERKS__)
+
 #define AT_ADDRESS(x) : (x)
+
 #else
+
 #define AT_ADDRESS(x)
+
 #endif
```
Comment (ribbanya) [inline include/dolphin/types.h:78]: @r-burns This, right?
```suggestion
#define AT_ADDRESS(x) = FIXEDADDR(x)
#elif defined(__MWERKS__)
#define AT_ADDRESS(x) : (x)
#else
#define AT_ADDRESS(x)
#endif
```

File: include/string.h (hunk 6 above)
Comment (ribbanya) [reply include/string.h:6]: Was a duplicate of something, either `Runtime/__mem.h` or `cstring.h`.

File: src/melee/ft/chara/ftCLink/ftclink.h (hunk 5 above)
Comment (ribbanya) [reply src/melee/ft/chara/ftCLink/ftclink.h:7]: Yeah, they are the same struct. Same applies to all clone characters it seems like (and Nana).

File: src/dolphin/os/OSExi.c (hunk 4 above)
Comment (ribbanya) [reply src/dolphin/os/OSExi.c:127]: Did you compile this? I think I tried it.

File: src/dolphin/card/CARDCreate.c (hunk 2 above)
Comment (ribbanya) [reply src/dolphin/card/CARDCreate.c:127]: Yeah I know, just seemed out of scope for this PR.

File: include/dolphin/types.h (hunk 7 above)
Comment (r-burns) [reply include/dolphin/types.h:78]: yeah lol

File: include/string.h (hunk 6 above)
Comment (r-burns) [reply include/string.h:6]: well I think it belongs here, it's C standard and <cstring> is a C++ thing

File: src/melee/ft/chara/ftCLink/ftclink.h (hunk 5 above)
Comment (r-burns) [reply src/melee/ft/chara/ftCLink/ftclink.h:7]: Interesting, thanks

File: src/dolphin/card/CARDCreate.c (hunk 2 above)
Comment (r-burns) [reply src/dolphin/card/CARDCreate.c:127]: Ok that's fair

File: src/dolphin/os/OSExi.c (hunk 4 above)
Comment (r-burns) [reply src/dolphin/os/OSExi.c:127]: Yeah it works

File: include/string.h (hunk 6 above)
Comment (ribbanya) [reply include/string.h:6]: Then our splits are wrong and it needs to be addressed at the TU level, not the header level.

File: src/dolphin/os/OSExi.c (hunk 4 above)
Comment (ribbanya) [reply src/dolphin/os/OSExi.c:127]: It also just does nothing as far as I can tell...

File: include/string.h (hunk 6 above)
Comment (r-burns) [reply include/string.h:6]: Nah where it's implemented is arbitrary, it just needs to be made available from this header. So just have this header include `<Runtime/__mem.h>` and I think we're good.

Comment (ribbanya) [reply include/string.h:6]: > So just have this header include `<Runtime/__mem.h>` and I think we're good.

Okay.

File: src/dolphin/os/OSExi.c (hunk 4 above)
Comment (r-burns) [reply src/dolphin/os/OSExi.c:127]: It offsets buf by xLen bytes for the next iteration.

Comment (ribbanya) [reply src/dolphin/os/OSExi.c:127]: Oh right, duh.

**Review by r-burns — APPROVED** [review]

**Review by r-burns — APPROVED** [review]

**Pull request merged**